python/cugraph/structure/number_map.py

```diff
@@ -0,0 +1,716 @@
+                    columns=dict(zip(src_col_names, self.col_names)),
+                    copy=False,
+                )
+            )
+
+            if dst_col_names is not None:
+                tmp_dst = (
+                    df[dst_col_names]
+                    .groupby(dst_col_names)
+                    .count()
+                    .reset_index()
+                )
+                for newname, oldname in zip(self.col_names, dst_col_names):
+                    self.df[newname] = tmp[newname].append(tmp_dst[oldname])
+
+            self.numbered = False
+
+        def compute(self):
+            if not self.numbered:
+                tmp = self.df.groupby(self.col_names).count().reset_index()
+                tmp["id"] = tmp.index.astype(np.int32)
+                self.df = tmp
+                self.numbered = True
+
+        def to_vertex_id(self, df, col_names):
+            tmp_df = df[col_names].rename(
+                columns=dict(zip(col_names, self.col_names)), copy=False
+            )
+            tmp_df["index"] = tmp_df.index
+            return (
+                tmp_df.merge(self.df, on=self.col_names, how="left")
+                .sort_values("index")
+                .drop(["index"])
+                .reset_index()["id"]
+            )
+
+        def add_vertex_id(self, df, id_column_name, col_names, drop):
+            ret = None
+            if col_names is None:
+                ret = df.merge(self.df, on=self.col_names, how="left").rename(
+                    columns={"id": id_column_name}, copy=False
+                )
+            elif col_names == self.col_names:
+                ret = df.merge(self.df, on=self.col_names, how="left").rename(
+                    columns={"id": id_column_name}, copy=False
+                )
+            else:
+                ret = (
+                    df.merge(
+                        self.df,
+                        left_on=col_names,
+                        right_on=self.col_names,
+                        how="left",
+                    )
+                    .drop(self.col_names)
+                    .rename(columns={"id": id_column_name}, copy=False)
+                )
+
+            if drop:
+                return ret.drop(col_names)
+
+            return ret
+
+        def from_vertex_id(
+            self, df, internal_column_name, external_column_names
+        ):
+            tmp_df = df.merge(
+                self.df,
+                left_on=internal_column_name,
+                right_on="id",
+                how="left",
+            )
+            if internal_column_name != "id":
+                tmp_df = tmp_df.drop(["id"])
+            if external_column_names is None:
+                return tmp_df
+            else:
+                return tmp_df.rename(
+                    columns=dict(zip(self.col_names, external_column_names)),
+                    copy=False,
+                )
+
+    class MultiGPU:
+        def extract_vertices(
```

Comment:
I am still digging into the details and thinking about potential implementation issues,

but if we wish to generate a CSR-ish graph, similar to the single-GPU case, if we can generate a list of (vid, local-out-degree) pairs here, and hash vertex id pairs to compute the target partition (I think you're already doing this), and collect (vid, local-out-degree) pairs in the target partition. In the target partition, we can reduce to compute global-out-degree, and sort before renumbering. If we can do this with low performance overhead (local sorting within a partition might be the most expensive part, but this is cheaper than global sorting), this will make remaining steps much easier (and partitioning becomes really straightforward as well).

Reply:
So at the end of renumbering, a NumberMap object will hold partitioning of [0, V) (assuming V vertices) to P processes, and within each partition (approximately sizeof V/P assuming good hash function), vertices are either sorted by out-degree (if we wish to generate a CSR-ish graph) or in-degree (if we wish to generate a CSC-ish graph).

I think the additional overhead for maintaining this sorted order within a partition can be pretty moderate unless E/V is really small; shuffling edges based on partitioning to build a CSR-ish graph representation (something we cannot avoid) will be more expensive than this...

Reply:
Question is, do we need to address these performance optimizations in this PR, or can they be added as we advance a bit further down the multi-GPU path?

This PR makes a significant API change to the Single GPU implementation and affects every current algorithm.  The longer we keep this open the more changes get made to the algorithms that have to be retrofitted to the new API.  My inclination would be to defer these optimizations to later in the MG development.

Reply:
If you want to merge this PR early, you can do this in a later PR, I think we can actually do things in two more PRs, one additional PR doing this optimization without actually writing high-performance C++ code (so we can concurrently working on following steps relying on this), and the last PR trying to cut the graph preparation time as much as possible.